desc/vmec_utils.py

```diff
@@ -469,3 +487,509 @@ def vmec_boundary_subspace(eq, RBC=None, ZBS=None, RBS=None, ZBC=None):  # noqa:
+
+    timer.start("Boozer Transform")
+
+    B_mn = np.array([[]])
+    R_mn = np.array([[]])
+    Z_mn = np.array([[]])
+    Nu_mn = np.array([[]])
+    Sqrt_g_B_mn = np.array([[]])
+
+    vol_grid = LinearGrid(M=2 * M_booz, N=2 * N_booz, NFP=eq.NFP, rho=r_half)
+    # precompute the needed data for the boozer surface computations
+    # (except those which have to be computed on a single surface only,
+    # like theta_B, zeta_B or nu)
+    keys = ["|B|", "R", "Z", "sqrt(g)", "rho", "psi_r", "lambda", "B_zeta", "B_theta"]
+    data_vol = eq.compute(
+        keys,
+        grid=vol_grid,
+    )
+
+    # FourierZernike fit the B_theta and B_zeta, then at each surface
+    # just evaluate the fit at the rho to get the B_theta_mn and B_zeta_mn
+    # so this we only only do a single fit
+    transform = Transform(
+        grid=vol_grid,
+        basis=FourierZernikeBasis(
+            L=2 * eq.L, M=M_booz - 1, N=N_booz, sym="cos", NFP=eq.NFP
+        ),
+        build_pinv=True,
+    )
+    B_zeta_lmn = transform.fit(data_vol["B_zeta"])
+    B_theta_lmn = transform.fit(data_vol["B_theta"])
+    # compute the fourier coeffs of the B_zeta and B_theta at each rho
+    _, _, B_zeta_mn = zernike_to_fourier(
+        B_zeta_lmn, transform.basis, rho=r_half, return_full=False
+    )
+    _, _, B_theta_mn = zernike_to_fourier(
+        B_theta_lmn, transform.basis, rho=r_half, return_full=False
+    )
+
+    for i, r in enumerate(r_half):
+        if verbose > 0:
+            printstring = f"Calculating Surf {i} at rho={r:1.3f}"
+            print("#" * len(printstring) + "\n" + printstring + "\n")
+
+        data = {}
+        for key in keys:
+            # populate the pre-computed data for this surface
+            data[key] = data_vol[key][np.where(vol_grid.nodes[:, 0] == r)]
+
+        # are of shape [num_rhos,num_modes], so index into the first axis
+        data["B_zeta_mn"] = B_zeta_mn[i, :]
+        data["B_theta_mn"] = B_theta_mn[i, :]
+
+        ## calculate boozer transform for this surface
+        grid = LinearGrid(M=2 * M_booz, N=2 * N_booz, NFP=eq.NFP, rho=np.array(r))
+        # cos symmetric terms
+        transforms = get_transforms(
+            ["|B|_mn", "sqrt(g)_B_mn"],
+            obj=eq,
+            grid=grid,
+            M_booz=M_booz - 1,
+            N_booz=N_booz,
+        )
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            data = eq.compute(
+                ["|B|_mn", "R_mn", "Z_mn", "nu_mn", "sqrt(g)_B_mn", "psi_r"],
+                grid=grid,
+                transforms=transforms,
+                data=data,
+            )
+        if eq.sym:
+            # Z and nu are sin-symmetric, but the transforms used
+            # before are cos symmetric.
+            # We want to use the data that needed cos sym transforms,
+            # but remove the incorrectly calculated Z_mn,nu_mn from the
+            # dictionary first
+            data.pop("nu_mn")
```

Comment:
wouldn't it be better to just not compute those qtys in the first place on the wrong grid?

Reply:
they are needed for the other cosine quantities, however they are not needed for the sin quantities. and if you are not stellarator symmetric, then the grid and basis is the same for everything and this block is not entered.

I could change it tho to be more efficient